File: experimental/ip-geofeed/internal/geofeed_validation/validator.go
package geofeed_validation

import (
	"encoding/json"
	"fmt"
	"ip-geofeed/internal/parser"
	"net"
	"os"
	"regexp"
	"strings"
	"time"
)

var (
	placeholderValues = []string{"undefined", "please select", "null", "n/a", "tbd", "unknown"}
	abbreviatedValues = []string{"la", "frft", "sin01", "lhr", "sin", "maa"}
)

// Country represents an ISO 3166-1 country
type Country struct {
	Alpha2       string `json:"alpha_2"`
	Alpha3       string `json:"alpha_3"`
	Flag         string `json:"flag"`
	Name         string `json:"name"`
	Numeric      string `json:"numeric,omitempty"`
	OfficialName string `json:"official_name,omitempty"`
}

// Region represents an ISO 3166-2 subdivision
type Region struct {
	Code string `json:"code"`
	Name string `json:"name"`
	Type string `json:"type"`
}

// ValidationContext holds all validation data
type ValidationContext struct {
	Countries        map[string]Country
	Regions          map[string]Region
	SmallTerritories map[string]bool
}

// Message represents a validation issue
type Message struct {
	ID      string
	Type    string // ERROR, WARNING, SUGGESTION
	Text    string
	Checked bool
}

// Entry represents a single CSV row with validation results
type Entry struct {
	parser.Row
	Status         string // ERROR, WARNING, SUGGESTION, OK
	Messages       []Message
	HasError       bool
	HasWarning     bool
	HasSuggestion  bool
	IPVersion      string
	DoNotGeolocate bool
	GeocodingHint  string
	Tunable        bool
	TunedEntry     Location
}

// Metadata represents summary information
type Metadata struct {
	InputFile            string
	Timestamp            int64
	TotalEntries         int
	IpV4Entries          int
	IpV6Entries          int
	InvalidEntries       int
	Errors               int
	Warnings             int
	Suggestions          int
	OK                   int
	CityLevelAccuracy    int
	RegionLevelAccuracy  int
	CountryLevelAccuracy int
	DoNotGeolocate       int
}

type Record struct {
	parser.Record
	ReportURL string
}

// Netname represents a grouping of geofeeds by netname
type Netname struct {
	Name     string
	Records  []Record
	TableURL string
}

// RIR represents a Regional Internet Registry with its associated netnames
type RIR struct {
	Name     string
	Netnames map[string]*Netname
}

// RIRCollection represents all Regional Internet Registries
type RIRCollection struct {
	RIRs map[string]*RIR
}

// LoadValidationData loads ISO and territorial data from JSON files
func LoadValidationData() (*ValidationContext, error) {
	ctx := &ValidationContext{
		Countries:        make(map[string]Country),
		Regions:          make(map[string]Region),
		SmallTerritories: make(map[string]bool),
	}

	// Load ISO 3166-1 countries
	countriesFile, err := os.ReadFile("internal/geofeed_validation/iso3166-1.json")
	if err != nil {
		return nil, fmt.Errorf("reading countries file: %w", err)
	}

	var countriesData struct {
		Countries []Country `json:"3166-1"`
	}
	if err := json.Unmarshal(countriesFile, &countriesData); err != nil {
		return nil, fmt.Errorf("parsing countries file: %w", err)
	}

	for _, country := range countriesData.Countries {
		ctx.Countries[country.Alpha2] = country
	}

	// Load ISO 3166-2 regions
	regionsFile, err := os.ReadFile("internal/geofeed_validation/iso3166-2.json")
	if err != nil {
		return nil, fmt.Errorf("reading regions file: %w", err)
	}

	var regionsData struct {
		Regions []Region `json:"3166-2"`
	}
	if err := json.Unmarshal(regionsFile, &regionsData); err != nil {
		return nil, fmt.Errorf("parsing regions file: %w", err)
	}

	for _, region := range regionsData.Regions {
		ctx.Regions[region.Code] = region
	}

	// Load small territories
	territoriesFile, err := os.ReadFile("internal/geofeed_validation/small-territories.json")
	if err != nil {
		return nil, fmt.Errorf("reading small territories file: %w", err)
	}

	var territories []string
	if err := json.Unmarshal(territoriesFile, &territories); err != nil {
		return nil, fmt.Errorf("parsing small territories file: %w", err)
	}

	for _, territory := range territories {
		ctx.SmallTerritories[territory] = true
	}

	return ctx, nil
}

// ValidateEntries validates a list of entries and populates their messages and status
func ValidateEntries(rows []parser.Row) ([]Entry, []Entry, error) {
	// Load validation data
	ctx, err := LoadValidationData()
	if err != nil {
		return nil, nil, fmt.Errorf("error loading validation data: %w", err)
	}
	entries, errEntries := GetEntriesFromServer(rows, ctx)

	// Validate each entry
	for i := range entries {
		ValidateEntry(&entries[i], ctx)
	}
	return entries, errEntries, nil
}

// // ValidateAndTuneEntries validates entries and then applies tuning recommendations
// func ValidateAndTuneEntries(rows []parser.Row) error {
// 	entries, err := ValidateEntries(rows)
// 	if err != nil {
// 		return err
// 	}

// 	// Tune entries based on validation results and geocoding hints
// 	TuneEntries(entries)

// 	return nil
// }

func GetMetadataFromEntries(entries []Entry, inputFile string, invalidEntries int) Metadata {
	metadata := Metadata{
		InputFile:      inputFile,
		Timestamp:      time.Now().UnixMilli(),
		InvalidEntries: invalidEntries,
	}

	for _, entry := range entries {
		metadata.TotalEntries++
		switch entry.IPVersion {
		case "IPv4":
			metadata.IpV4Entries++
		case "IPv6":
			metadata.IpV6Entries++
		}

		if entry.HasError {
			metadata.Errors++
		} else if entry.HasWarning {
			metadata.Warnings++
		} else if entry.HasSuggestion {
			metadata.Suggestions++
		} else {
			metadata.OK++
		}

		if entry.City != "" {
			metadata.CityLevelAccuracy++
		} else if entry.RegionCode != "" {
			metadata.RegionLevelAccuracy++
		} else if entry.CountryCode != "" {
			metadata.CountryLevelAccuracy++
		}

		if entry.DoNotGeolocate {
			metadata.DoNotGeolocate++
		}
	}

	return metadata
}

// ValidateEntry validates a single entry and populates its messages
func ValidateEntry(entry *Entry, ctx *ValidationContext) {
	// 1. IP Prefix validation
	ValidateIPPrefix(entry)

	// 2. Country code validation
	ValidateCountryCode(entry, ctx)

	// 3. Region code validation
	ValidateRegionCode(entry, ctx)

	// 4. City name validation
	ValidateCityName(entry, ctx)

	// 5. Postal code check (deprecated)
	ValidatePostalCode(entry)

	// 6. Tuning recommendations
	ProvideTuningRecommendations(entry, ctx)

	// Determine overall status
	if entry.HasError {
		entry.Status = "ERROR"
	} else if entry.HasWarning {
		entry.Status = "WARNING"
	} else if entry.HasSuggestion {
		entry.Status = "SUGGESTION"
	} else {
		entry.Status = "OK"
	}
}

// LoadRIRData organizes Records into a RIRCollection structure
// by grouping geofeeds by their RIR (Source) and then by netname
func LoadRIRData(publishers parser.Records) *RIRCollection {
	// Create a map to group geofeeds by RIR (Source)
	rirMap := make(map[string]*RIR)

	// Group geofeeds by RIR and then by netname
	for _, record := range publishers {

		if record.Source == "" || record.Netname == "" {
			continue // Skip entries without a source/RIR or netname
		}

		rirName := record.Source

		// Create RIR if it doesn't exist
		if _, exists := rirMap[rirName]; !exists {
			rirMap[rirName] = &RIR{
				Name:     rirName,
				Netnames: make(map[string]*Netname),
			}
		}

		rir := rirMap[rirName]
		netnameKey := record.Netname

		// Create Netname if it doesn't exist
		if _, exists := rir.Netnames[netnameKey]; !exists {
			rir.Netnames[netnameKey] = &Netname{
				Name:    netnameKey,
				Records: []Record{},
			}
		}

		// Add record to the netname
		rir.Netnames[netnameKey].Records = append(rir.Netnames[netnameKey].Records, Record{
			Record:    record,
			ReportURL: "", // This can be populated later with the actual report URL after validation
		})
	}

	return &RIRCollection{
		RIRs: rirMap,
	}
}

// sanitize removes unsafe filesystem characters
func sanitize(s string) string {
	s = strings.TrimSpace(s)
	reg := regexp.MustCompile(`[^a-zA-Z0-9\-_]+`)
	return reg.ReplaceAllString(s, "")
}

func GetSourceTablePath(source string) string {
	if source != "" {
		return strings.ToLower(sanitize(source))
	}
	return ""
}

func GetNetnameTablePath(netname string) string {
	if netname != "" {
		return sanitize(netname)
	}
	return ""
}

// ValidateIPPrefix validates the IP prefix format and properties
func ValidateIPPrefix(entry *Entry) {
	if entry.IPPrefix == "" {
		entry.AddStatusMessage(ErrIPPrefixEmpty)
		return
	}

	// Try to parse as CIDR
	_, ipNet, err := net.ParseCIDR(entry.IPPrefix)
	if err != nil {
		entry.AddStatusMessage(ErrIPPrefixInvalid)
		return
	}

	// Normalize to CIDR notation
	entry.IPPrefix = ipNet.String()
	if ipNet.IP.To4() != nil {
		entry.IPVersion = "IPv4"
	} else {
		entry.IPVersion = "IPv6"
	}

	// Check if it's a public address
	if IsPrivateAddress(ipNet.IP) {
		entry.AddStatusMessage(ErrIPPrefixNonPublic)
		return
	}

	// Check for overly large prefixes
	if entry.IPVersion == "IPv4" {
		ones, _ := ipNet.Mask.Size()
		if ones < 22 {
			entry.AddStatusMessage(SuggestIPv4PrefixLarge)
		}
	} else { // IPv6
		ones, _ := ipNet.Mask.Size()
		if ones < 64 {
			entry.AddStatusMessage(SuggestIPv6PrefixLarge)
		}
	}
}

// IsPrivateAddress checks if an IP is private or reserved
func IsPrivateAddress(ip net.IP) bool {
	if ip.IsLoopback() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() || ip.IsMulticast() {
		return true
	}

	// Check for private IPv4 ranges
	if ip.To4() != nil {
		privateRanges := []string{
			"10.0.0.0/8",
			"172.16.0.0/12",
			"192.168.0.0/16",
			"169.254.0.0/16",
			"127.0.0.0/8",
		}
		for _, cidr := range privateRanges {
			_, ipNet, _ := net.ParseCIDR(cidr)
			if ipNet.Contains(ip) {
				return true
			}
		}
	} else {
		// Check for private IPv6 ranges
		privateRanges := []string{
			"fc00::/7",  // Unique local addresses
			"fe80::/10", // Link-local
			"::1/128",   // Loopback
		}
		for _, cidr := range privateRanges {
			_, ipNet, _ := net.ParseCIDR(cidr)
			if ipNet.Contains(ip) {
				return true
			}
		}
	}

	return false
}

// ValidateCountryCode validates the ISO 3166-1 country code
func ValidateCountryCode(entry *Entry, ctx *ValidationContext) {
	if entry.CountryCode == "" {
		return // Empty is allowed
	}

	if _, exists := ctx.Countries[entry.CountryCode]; !exists {
		entry.AddStatusMessage(ErrCountryCodeInvalid)
	}
}

// ValidateRegionCode validates the ISO 3166-2 region code
func ValidateRegionCode(entry *Entry, ctx *ValidationContext) {
	if entry.RegionCode == "" {
		return // Empty is allowed
	}

	// Check format: COUNTRY-SUBDIVISION
	regionPattern := regexp.MustCompile(`^[A-Z]{2}-[A-Z0-9]{1,3}$`)
	if !regionPattern.MatchString(entry.RegionCode) {
		entry.AddStatusMessage(ErrRegionCodeFormat)
		return
	}

	// Check if region exists in ISO 3166-2
	if _, exists := ctx.Regions[entry.RegionCode]; !exists {
		entry.AddStatusMessage(ErrRegionCodeInvalid)
		return
	}

	// Check if region's country matches the country code
	if entry.CountryCode != "" {
		regionCountry := strings.Split(entry.RegionCode, "-")[0]
		if regionCountry != entry.CountryCode {
			entry.AddStatusMessage(ErrRegionCodeMismatch)
		}
	}
}

// ValidateCityName validates the city name for inconsistencies
func ValidateCityName(entry *Entry, ctx *ValidationContext) {
	if entry.City == "" {
		return // Empty is allowed
	}

	cityLower := strings.ToLower(strings.TrimSpace(entry.City))

	// Check for placeholder values
	for _, placeholder := range placeholderValues {
		if cityLower == placeholder {
			entry.AddStatusMessage(ErrCityPlaceholder)
			return
		}
	}

	// Check for abbreviated values
	for _, abbrev := range abbreviatedValues {
		if cityLower == abbrev {
			entry.AddStatusMessage(ErrCityAbbreviated)
			return
		}
	}

	// Check for inconsistent formatting (very basic heuristic)
	if strings.Contains(entry.City, "  ") || // Double spaces
		(strings.ToUpper(entry.City) == entry.City && len(entry.City) > 3) || // ALL CAPS
		regexp.MustCompile(`[A-Z][a-z]+[A-Z]`).MatchString(entry.City) { // HongKong style
		entry.AddStatusMessage(WarnCityFormattingBad)
	}
}

// ValidatePostalCode checks that postal codes are not included (deprecated by RFC 8805)
func ValidatePostalCode(entry *Entry) {
	if entry.PostalCode != "" {
		entry.AddStatusMessage(ErrPostalCodeDeprecated)
	}
}

func CheckForIsuues(entry *parser.Row, ctx *ValidationContext) bool {
	if entry.CountryCode == "" {
		return false
	}
	if _, exists := ctx.Countries[entry.CountryCode]; exists {
		return true
	}

	if _, exists := ctx.Regions[entry.RegionCode]; entry.RegionCode != "" && exists {
		return true
	}

	return false
}

File: experimental/ip-geofeed2/internal/geofeed_validation/validator.py
import ipaddress
import json
import os
import re
import time
from dataclasses import dataclass, field
from typing import Optional

from internal.parser.parser import Row, Record

from .errors import (
    ERROR_TYPE, WARNING_TYPE, SUGGEST_TYPE,
    ERR_IP_PREFIX_EMPTY, ERR_IP_PREFIX_INVALID, ERR_IP_PREFIX_NON_PUBLIC,
    SUGGEST_IPV4_PREFIX_LARGE, SUGGEST_IPV6_PREFIX_LARGE,
    ERR_COUNTRY_CODE_INVALID,
    ERR_REGION_CODE_FORMAT, ERR_REGION_CODE_INVALID, ERR_REGION_CODE_MISMATCH,
    ERR_CITY_PLACEHOLDER, ERR_CITY_ABBREVIATED, WARN_CITY_FORMATTING_BAD,
    ERR_POSTAL_CODE_DEPRECATED,
    SUGGEST_REGION_UNNECESSARY_SMALL_TERRITORY, SUGGEST_CITY_UNNECESSARY_SMALL_TERRITORY,
    SUGGEST_REGION_RECOMMENDED_WITH_CITY, SUGGEST_CONFIRM_DO_NOT_GEOLOCATE,
    ValidationError,
)

PLACEHOLDER_VALUES = ["undefined", "please select", "null", "n/a", "tbd", "unknown"]
ABBREVIATED_VALUES = ["la", "frft", "sin01", "lhr", "sin", "maa"]

# IPv4 private ranges
_IPV4_PRIVATE_RANGES = [
    ipaddress.ip_network("10.0.0.0/8"),
    ipaddress.ip_network("172.16.0.0/12"),
    ipaddress.ip_network("192.168.0.0/16"),
    ipaddress.ip_network("169.254.0.0/16"),
    ipaddress.ip_network("127.0.0.0/8"),
]

# IPv6 private ranges
_IPV6_PRIVATE_RANGES = [
    ipaddress.ip_network("fc00::/7"),
    ipaddress.ip_network("fe80::/10"),
    ipaddress.ip_network("::1/128"),
]


@dataclass
class Country:
    """Represents an ISO 3166-1 country."""
    alpha_2: str = ""
    alpha_3: str = ""
    flag: str = ""
    name: str = ""
    numeric: str = ""
    official_name: str = ""


@dataclass
class Region:
    """Represents an ISO 3166-2 subdivision."""
    code: str = ""
    name: str = ""
    type: str = ""


@dataclass
class ValidationContext:
    """Holds all validation data."""
    countries: dict[str, Country] = field(default_factory=dict)
    regions: dict[str, Region] = field(default_factory=dict)
    small_territories: dict[str, bool] = field(default_factory=dict)


@dataclass
class Message:
    """Represents a validation issue."""
    id: str = ""
    type: str = ""
    text: str = ""
    checked: bool = False


@dataclass
class Location:
    """Represents a geographic location match."""
    name: str = ""
    country_code: str = ""
    region_code: str = ""
    place_type: str = ""
    h3_cells: list[str] = field(default_factory=list)
    bounding_box: list[float] = field(default_factory=list)


@dataclass
class Entry:
    """Represents a single CSV row with validation results."""
    # From Row
    line: int = 0
    ip_prefix: str = ""
    country_code: str = ""
    region_code: str = ""
    city: str = ""
    postal_code: str = ""
    # Validation fields
    status: str = ""
    messages: list[Message] = field(default_factory=list)
    has_error: bool = False
    has_warning: bool = False
    has_suggestion: bool = False
    ip_version: str = ""
    do_not_geolocate: bool = False
    geocoding_hint: str = ""
    tunable: bool = False
    tuned_entry: Optional[Location] = None

    def add_status_message(self, msg: ValidationError):
        self.messages.append(Message(
            id=msg.id,
            type=msg.type,
            text=msg.message,
            checked=msg.tunable,
        ))
        self.tunable = self.tunable or msg.tunable
        if msg.type == ERROR_TYPE:
            self.has_error = True
        elif msg.type == WARNING_TYPE:
            self.has_warning = True
        elif msg.type == SUGGEST_TYPE:
            self.has_suggestion = True


@dataclass
class Metadata:
    """Represents summary information."""
    input_file: str = ""
    timestamp: int = 0
    total_entries: int = 0
    ipv4_entries: int = 0
    ipv6_entries: int = 0
    invalid_entries: int = 0
    errors: int = 0
    warnings: int = 0
    suggestions: int = 0
    ok: int = 0
    city_level_accuracy: int = 0
    region_level_accuracy: int = 0
    country_level_accuracy: int = 0
    do_not_geolocate: int = 0


@dataclass
class ValidatorRecord:
    """Wraps a parser Record with a report URL."""
    record: Record = field(default_factory=Record)
    report_url: str = ""


@dataclass
class Netname:
    """Represents a grouping of geofeeds by netname."""
    name: str = ""
    records: list[ValidatorRecord] = field(default_factory=list)
    table_url: str = ""


@dataclass
class RIR:
    """Represents a Regional Internet Registry with its associated netnames."""
    name: str = ""
    netnames: dict[str, Netname] = field(default_factory=dict)


@dataclass
class RIRCollection:
    """Represents all Regional Internet Registries."""
    rirs: dict[str, RIR] = field(default_factory=dict)


def load_validation_data() -> ValidationContext:
    """Loads ISO and territorial data from JSON files."""
    ctx = ValidationContext()

    # Load ISO 3166-1 countries
    with open("internal/geofeed_validation/iso3166-1.json", "r") as f:
        countries_data = json.load(f)
    for c in countries_data.get("3166-1", []):
        country = Country(
            alpha_2=c.get("alpha_2", ""),
            alpha_3=c.get("alpha_3", ""),
            flag=c.get("flag", ""),
            name=c.get("name", ""),
            numeric=c.get("numeric", ""),
            official_name=c.get("official_name", ""),
        )
        ctx.countries[country.alpha_2] = country

    # Load ISO 3166-2 regions
    with open("internal/geofeed_validation/iso3166-2.json", "r") as f:
        regions_data = json.load(f)
    for r in regions_data.get("3166-2", []):
        region = Region(
            code=r.get("code", ""),
            name=r.get("name", ""),
            type=r.get("type", ""),
        )
        ctx.regions[region.code] = region

    # Load small territories
    with open("internal/geofeed_validation/small-territories.json", "r") as f:
        territories = json.load(f)
    for t in territories:
        ctx.small_territories[t] = True

    return ctx


def validate_entries(rows: list[Row]) -> tuple[list[Entry], list[Entry]]:
    """Validates a list of entries and populates their messages and status."""
    from .tuner import get_entries_from_server

    ctx = load_validation_data()
    entries, err_entries = get_entries_from_server(rows, ctx)

    for entry in entries:
        validate_entry(entry, ctx)

    return entries, err_entries


def get_metadata_from_entries(entries: list[Entry], input_file: str, invalid_entries: int) -> Metadata:
    """Computes metadata summary from validated entries."""
    metadata = Metadata(
        input_file=input_file,
        timestamp=int(time.time() * 1000),
        invalid_entries=invalid_entries,
    )

    for entry in entries:
        metadata.total_entries += 1
        if entry.ip_version == "IPv4":
            metadata.ipv4_entries += 1
        elif entry.ip_version == "IPv6":
            metadata.ipv6_entries += 1

        if entry.has_error:
            metadata.errors += 1
        elif entry.has_warning:
            metadata.warnings += 1
        elif entry.has_suggestion:
            metadata.suggestions += 1
        else:
            metadata.ok += 1

        if entry.city:
            metadata.city_level_accuracy += 1
        elif entry.region_code:
            metadata.region_level_accuracy += 1
        elif entry.country_code:
            metadata.country_level_accuracy += 1

        if entry.do_not_geolocate:
            metadata.do_not_geolocate += 1

    return metadata


def validate_entry(entry: Entry, ctx: ValidationContext):
    """Validates a single entry and populates its messages."""
    validate_ip_prefix(entry)
    validate_country_code(entry, ctx)
    validate_region_code(entry, ctx)
    validate_city_name(entry, ctx)
    validate_postal_code(entry)
    provide_tuning_recommendations(entry, ctx)

    if entry.has_error:
        entry.status = "ERROR"
    elif entry.has_warning:
        entry.status = "WARNING"
    elif entry.has_suggestion:
        entry.status = "SUGGESTION"
    else:
        entry.status = "OK"


def load_rir_data(publishers: list[Record]) -> RIRCollection:
    """Organizes Records into a RIRCollection structure."""
    rir_map: dict[str, RIR] = {}

    for record in publishers:
        if not record.source or not record.netname:
            continue

        rir_name = record.source

        if rir_name not in rir_map:
            rir_map[rir_name] = RIR(name=rir_name)

        rir = rir_map[rir_name]
        netname_key = record.netname

        if netname_key not in rir.netnames:
            rir.netnames[netname_key] = Netname(name=netname_key)

        rir.netnames[netname_key].records.append(ValidatorRecord(
            record=record,
            report_url="",
        ))

    return RIRCollection(rirs=rir_map)


def _sanitize(s: str) -> str:
    """Removes unsafe filesystem characters."""
    s = s.strip()
    return re.sub(r"[^a-zA-Z0-9\-_]+", "", s)


def get_source_table_path(source: str) -> str:
    if source:
        return _sanitize(source).lower()
    return ""


def get_netname_table_path(netname: str) -> str:
    if netname:
        return _sanitize(netname)
    return ""


def validate_ip_prefix(entry: Entry):
    """Validates the IP prefix format and properties."""
    if not entry.ip_prefix:
        entry.add_status_message(ERR_IP_PREFIX_EMPTY)
        return

    try:
        network = ipaddress.ip_network(entry.ip_prefix, strict=False)
    except ValueError:
        entry.add_status_message(ERR_IP_PREFIX_INVALID)
        return

    entry.ip_prefix = str(network)

    if isinstance(network, ipaddress.IPv4Network):
        entry.ip_version = "IPv4"
    else:
        entry.ip_version = "IPv6"

    if is_private_address(network.network_address):
        entry.add_status_message(ERR_IP_PREFIX_NON_PUBLIC)
        return

    if entry.ip_version == "IPv4":
        if network.prefixlen < 22:
            entry.add_status_message(SUGGEST_IPV4_PREFIX_LARGE)
    else:
        if network.prefixlen < 64:
            entry.add_status_message(SUGGEST_IPV6_PREFIX_LARGE)


def is_private_address(ip: ipaddress.IPv4Address | ipaddress.IPv6Address) -> bool:
    """Checks if an IP is private or reserved."""
    if ip.is_loopback or ip.is_link_local or ip.is_multicast:
        return True

    if isinstance(ip, ipaddress.IPv4Address):
        for net in _IPV4_PRIVATE_RANGES:
            if ip in net:
                return True
    else:
        for net in _IPV6_PRIVATE_RANGES:
            if ip in net:
                return True

    return False


def validate_country_code(entry: Entry, ctx: ValidationContext):
    """Validates the ISO 3166-1 country code."""
    if not entry.country_code:
        return
    if entry.country_code not in ctx.countries:
        entry.add_status_message(ERR_COUNTRY_CODE_INVALID)


def validate_region_code(entry: Entry, ctx: ValidationContext):
    """Validates the ISO 3166-2 region code."""
    if not entry.region_code:
        return

    region_pattern = re.compile(r"^[A-Z]{2}-[A-Z0-9]{1,3}$")
    if not region_pattern.match(entry.region_code):
        entry.add_status_message(ERR_REGION_CODE_FORMAT)
        return

    if entry.region_code not in ctx.regions:
        entry.add_status_message(ERR_REGION_CODE_INVALID)
        return

    if entry.country_code:
        region_country = entry.region_code.split("-")[0]
        if region_country != entry.country_code:
            entry.add_status_message(ERR_REGION_CODE_MISMATCH)


def validate_city_name(entry: Entry, ctx: ValidationContext):
    """Validates the city name for inconsistencies."""
    if not entry.city:
        return

    city_lower = entry.city.strip().lower()

    for placeholder in PLACEHOLDER_VALUES:
        if city_lower == placeholder:
            entry.add_status_message(ERR_CITY_PLACEHOLDER)
            return

    for abbrev in ABBREVIATED_VALUES:
        if city_lower == abbrev:
            entry.add_status_message(ERR_CITY_ABBREVIATED)
            return

    if ("  " in entry.city or
            (entry.city.upper() == entry.city and len(entry.city) > 3) or
            re.search(r"[A-Z][a-z]+[A-Z]", entry.city)):
        entry.add_status_message(WARN_CITY_FORMATTING_BAD)


def validate_postal_code(entry: Entry):
    """Checks that postal codes are not included (deprecated by RFC 8805)."""
    if entry.postal_code:
        entry.add_status_message(ERR_POSTAL_CODE_DEPRECATED)


def provide_tuning_recommendations(entry: Entry, ctx: ValidationContext):
    """Provides suggestions for optimizing geofeed entries."""
    is_small_territory = ctx.small_territories.get(entry.country_code, False)

    if is_small_territory and entry.region_code:
        entry.add_status_message(SUGGEST_REGION_UNNECESSARY_SMALL_TERRITORY)

    if is_small_territory and entry.city:
        entry.add_status_message(SUGGEST_CITY_UNNECESSARY_SMALL_TERRITORY)

    if not is_small_territory and entry.city and not entry.region_code:
        entry.add_status_message(SUGGEST_REGION_RECOMMENDED_WITH_CITY)

    if not entry.country_code and not entry.region_code and not entry.city:
        entry.add_status_message(SUGGEST_CONFIRM_DO_NOT_GEOLOCATE)
        entry.do_not_geolocate = True


def check_for_issues(country_code: str, region_code: str, ctx: ValidationContext) -> bool:
    """Checks if country/region codes have issues."""
    if country_code and country_code not in ctx.countries:
        return True
    if region_code and region_code not in ctx.regions:
        return True
    return False
